index.js

```diff
@@ -98,34 +105,63 @@ class Dynastar {
 
     const query = key ? this.model.query(key) : this.model.scan();
     if (fields) query.attributes(fields);
-    // These models are weird and dont even have proper getters so we have to
+
+    return this.findAllQuery(query, callback);
+  }
+
+  /**
+   * Safely runs a query (or table scan).
+   * This is useful when you need to query a secondary indexes and want our default item parsing logic.
+   *
+   * @function findAllQuery
+   * @param {Object} query A DynamoDB model query or scan object to execute
+   * @param {Function} [callback] Continuation function when finished
+   * @returns {Stream} Stream with results
+   */
+  findAllQuery(query, callback) {
+    if (callback) {
+      callback = once(callback);
+    }
+
+    // These models are weird and don't even have proper getters so we have to
     // just use it as raw json so we can expect the object to have
     // properties
-    const stream = query.loadAll().exec().pipe(through.obj(function (res, enc, cb) {
+    const rawStream = query.loadAll().exec();
+    const stream = rawStream.pipe(through.obj(function (res, enc, cb) {
       res = res.Items || res;
       for (const d of res) {
         this.push(d && d.toJSON());
       }
       cb();
     }));
-    if (callback) return stream.pipe(ls.obj(callback));
-    return stream;
 
+    if (callback) {
+      const lsStream = stream.pipe(ls.obj(callback));
+      rawStream.on('error', callback);
+      stream.on('error', callback);
+      lsStream.on('error', callback);
+      return lsStream;
+    }
+
+    return stream;
```

Comment:
I've fixed the callback variant here, but I assume something needs to happen on the stream side as well though.

I assume we'd want to create a stream that we write to and return and proxy all the calls to that stream? Is that right?

Reply:
Found a fix and a test that seems to work. See 429af56